gh-108277: Add wrapper for timerfd_create, timerfd_settime, and timerfd_gettime to os module

m-tmatma opened the pull request:
Add wrapper for timerfd system calls.

<!-- gh-issue-number: gh-108277 -->
* Issue: gh-108277
<!-- /gh-issue-number -->

Note: local test

```
python3 Tools/clinic/clinic.py -f Modules/posixmodule.c &&  \
autoconf -i && \
./configure && \
make regen-configure && \
make -j4 regen-all   && \
make regen-stdlib-module-names &&  \
make test TESTOPTS="-v test_os" && \
make -C Doc venv && make -C Doc html
```

<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--108382.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

- https://cpython-previews--108382.org.readthedocs.build/en/108382/library/os.html
- https://cpython-previews--108382.org.readthedocs.build/en/108382/whatsnew/3.13.html

m-tmatma committed:
re-run Tools/clinic/clinic.py

m-tmatma committed:
gh-108277: Add wrapper functions for timerfd_create, timerfd_settime, and timerfd_gettime to 'os' module.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

m-tmatma committed:
fix lint error

serhiy-storchaka commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
 
+.. function:: timerfd_create(clockid, flags)
+
+   Create and return an timerfd file descriptor. The file descriptors supports
```
Should not it be `a timer file descriptor`?

m-tmatma replied:
@serhiy-storchaka 
Fixed.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10015,6 +10020,174 @@ os_times_impl(PyObject *module)
+
+    clockid: int
+    flags: int
+
```
Make parameters positional-only?
```suggestion
    /

```

m-tmatma replied:
@serhiy-storchaka 
Updated.

serhiy-storchaka reviewed (commented):
In general LGTM, but we must first discuss what exactly we want to get.

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -16634,6 +16825,16 @@ static const struct have_function {
+#endif
+#ifdef HAVE_TIMERFD_GETTIME
+    {"HAVE_TIMERFD_GETTIME", NULL},
+#endif
```
It seems to me these two defines are not used to guard their respective functions. If you only need to check for `timerfd_create` to detect this feature, you can remove the `timerfd_settime` and `timerfd_gettime` checks from configure.ac; if not, you can add preprocessor guards for those respective functions in posixmodule.c.

erlend-aasland replied:
Alternatively, you could consolidate all of them in a single `HAVE_TIMEFD_API` define (in configure.ac), and simply use that as a guard in posixmodule.c.

erlend-aasland replied:
... or just use the `HAVE_SYS_TIMERFD_H` guard.

m-tmatma replied:
@erlend-aasland 
Fixed.

m-tmatma committed:
fix typo in doc

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

m-tmatma committed:
re-run Tools/clinic/clinic.py

m-tmatma committed:
define HAVE_TIMERFD_API only

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   ``it_value`` in ``struct itimerspec``. They are both in seconds unit and types are double.
+
+   ``interval`` is calculated like ``it_interval.tv_sec + it_interval.tv_nsec * 1e-9``,
+   ``value`` is calculated like ``it_interval.tv_sec + it_interval.tv_nsec * 1e-9`` internally.
```
The params should be marked up as `*param_name*`, the C struct members with ```:c:member:`!member` ```, and the C structs with ```:c:struct:`!struct s` ```.

I guess we could use `c:expr:` for the calculations, but I'm not sure about the Sphinx magic that happens with `c:expr:`. @AA-Turner would know.

Please also document the return value.

AA-Turner replied:
`:c:expr:` adds cross references to everything it can and leaves the rest with literal formatting (similar to using ``` `` ```). `c:texpr:` can be used to render it as normal text instead.

Reference: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#inline-expressions-and-types

A

m-tmatma replied:
@AA-Turner @erlend-aasland 
Fixed.

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   ``it_value`` in ``struct itimerspec``. They are both in nano second unit and types are long long.
+
+   ``interval`` is calculated like ``it_interval.tv_sec * 10**9 + it_interval.tv_nsec``,
+   ``value`` is calculated like ``it_interval.tv_sec * 10**9 + it_interval.tv_nsec`` internally.
```
Ditto

m-tmatma replied:
@erlend-aasland 
Fixed.

m-tmatma committed:
update What's New file

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+
+.. function:: timerfd_gettime(fd)
+
+   Returns a tuple ``interval`` which and ``value`` which is the time until next expiration.
```
Please use the imperative mood when writing reference text.

```suggestion
   Return a tuple ``interval`` which and ``value`` which is the time until next expiration.
```

Please make it clearer that the return value is a **two-item** tuple `(interval, value)`.

erlend-aasland replied:
Ditto for `timerfd_gettime_ns`.

m-tmatma replied:
@erlend-aasland 
Fixed.

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+
+   .. availability:: Linux >= 2.6.27 with glibc >= 2.8
+
+   .. versionadded:: 3.13
```
You can collect these constants under one entry. See for example the `SEEK_` or `PRIO_` constants.

m-tmatma replied:
@erlend-aasland 
Fixed.

m-tmatma committed:
Update document

m-tmatma committed:
fix document

m-tmatma committed:
fix document

m-tmatma committed:
fix comment

m-tmatma committed:
fix document

m-tmatma committed:
fix lint error

m-tmatma committed:
fix comment

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,164 @@ features:
+   .. c:var:: double interval;
+   .. c:var:: double value;
+
+   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9;`
```
```suggestion
   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9`
```

m-tmatma replied:
@AA-Turner 
Fixed.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,164 @@ features:
+
+   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9;`
+
+   :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9;`
```
```suggestion
   :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9`
```

m-tmatma replied:
@AA-Turner
Fixed.

AA-Turner committed:
Fix Sphinx C expr parsing (no semicolons)

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,164 @@ features:
+
+.. function:: timerfd_settime(fd, flags, interval, value)
+
+   Start/stop timer for an timer file descriptor.
```
```suggestion
   Start/stop timer for a timer file descriptor.
```

erlend-aasland commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,164 @@ features:
+.. function:: timerfd_settime(fd, flags, interval, value)
+
+   Start/stop timer for an timer file descriptor.
+   See man page :manpage:`timerfd_settime(2)` for more information.
```
@AA-Turner, what do you think of using `.. sealso::` for the manpage links?

m-tmatma replied:
@erlend-aasland 
Do you mean this?
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso
@AA-Turner

erlend-aasland replied:
Yes

m-tmatma replied:
@erlend-aasland
Fixed.
@AA-Turner

erlend-aasland commented on Lib/test/test_os.py:
```diff
@@ -3915,6 +3915,231 @@ def test_eventfd_select(self):
+        # 1st call
+        interval2, value2 = os.timerfd_settime(fd, 0, interval, value)
+        self.assertLessEqual(interval2, limit_error)
+        self.assertLessEqual(value2, limit_error)
```
Can you use `self.assertAlmostEqual` for these asserts?

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual

m-tmatma replied:
@erlend-aasland 
Fixed.

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+        2,
+        PyFloat_FromDouble((double)(curr_value->it_interval.tv_sec) + (double)curr_value->it_interval.tv_nsec * 1e-9),
+        PyFloat_FromDouble((double)(curr_value->it_value.tv_sec) + (double)curr_value->it_value.tv_nsec * 1e-9)
+    );
```
You'll need to add error handling and dereference temporary objects:

```suggestion
    double _interval = (double)(curr_value->it_interval.tv_sec) +
                       (double)curr_value->it_interval.tv_nsec * 1e-9;
    PyObject *interval = PyFloat_FromDouble(_inverval);
    if (interval == NULL) {
        return NULL;
    }
    double _value = (double)(curr_value->it_value.tv_sec) +
                    (double)curr_value->it_value.tv_nsec * 1e-9;
    PyObject *value = PyFloat_FromDouble(_value);
    if (value == NULL) {
        Py_DECREF(interval);
        return NULL;
    }
    PyObject *tuple = PyTuple_Pack(2, interval, value);
    Py_DECREF(interval);
    Py_DECREF(value);
    return tuple;
```

Ditto for similar functions.

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+    fd: fildes
+    /
+
+Read timerfd value in second.
```
```suggestion
Return timerfd value in seconds.
```

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+    it_value: double = 0.0
+    /
+
+Write timerfd value in second.
```
Missing s?

```suggestion
Set timerfd value in seconds.
```

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+    fd: fildes
+    /
+
+Read timerfd value in ns
```
```suggestion
Return timerfd value in nanoseconds.
```

m-tmatma replied:
@erlend-aasland 
Fixed.

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+    it_value_ns: long_long = 0
+    /
+
+Write timerfd value.
```
```suggestion
Set timerfd value in nanoseconds.
```

m-tmatma replied:
@erlend-aasland 
Fixed.

erlend-aasland commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,175 @@ os_times_impl(PyObject *module)
+    flags: int
+    /
+
+Creates and returns an timerfd notification file descriptor.
```
```suggestion
Create and return a timerfd notification file descriptor.
```

m-tmatma replied:
@erlend-aasland 
Fixed.

erlend-aasland commented on Modules/posixmodule.c:
Perhaps it would make sense to add param docstrings to all the new functions.

m-tmatma replied:
@erlend-aasland 
Added comment for parameters.
Is it ok?

erlend-aasland replied:
Sorry, I should have been more specific; my bad! :(

What I meant was param docstrings in the Argument Clinic input:

```
/*[clinic input]
os.timerfd_create
    clockid: int
        This is the param docstring for clockid
    flags: int
        This is the param docstring for flags;
        they can be multi-line!
    /

This is the method docstring. Note that the param docstrings
will be embedded into the rendered docstring by Argument Clinic.
```

m-tmatma replied:
@erlend-aasland 
Updated.

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
fix compile error

m-tmatma committed:
update hash

m-tmatma committed:
use self.assertAlmostEqual

m-tmatma committed:
Merge remote-tracking branch 'upstream/main' into feature/os.timerfd_xxx

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

m-tmatma committed:
add document for parameters

m-tmatma committed:
add param comment

m-tmatma committed:
use seealso

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+os.timerfd_create
+
+    clockid: int
+        CLOCK_REALTIME or CLOCK_MONOTONIC. CLOCK_REALTIME.
```
This is unclear.

m-tmatma replied:
@AA-Turner 
Added comments.

AA-Turner committed:
Merge branch 'main' into feature/os.timerfd_xxx

# Conflicts:
#	Modules/clinic/posixmodule.c.h

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+    clockid: int
+        CLOCK_REALTIME or CLOCK_MONOTONIC. CLOCK_REALTIME.
+    flags: int
+        0 or  a bit mask of TFD_NONBLOCK or TFD_CLOEXEC.
```
I think *flags* should default to 0, and perhaps should also be keyword-only. At the least, it should be positional-or-keyword.

m-tmatma replied:
@AA-Turner 
Changed to keyword only except for fd.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+
+
+/*[clinic input]
+os.timerfd_gettime
```
Very minor but I found it odd that gettime came before settime in the implementation, when they're reversed in the documentation and the man page!

m-tmatma replied:
@AA-Turner 
Reordered the definition. made them to settime first and gettime second

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+os.timerfd_gettime
+
+    fd: fildes
+        a file descriptor for timer.
```
You could be a bit more descriptive here

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+    fd: fildes
+        a file descriptor for timer.
+    flags: int
+        0 or  a bit mask of TFD_TIMER_ABSTIME or TFD_TIMER_CANCEL_ON_SET.
```
I think *flags* should default to 0, and perhaps should also be keyword-only. At the least, it should be positional-or-keyword.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+        it_interval is interval for timer in seconds.
+    it_value: double = 0.0
+        it_value is initial expiration timing in seconds. It could be absolute time or relative time
+        based on TFD_TIMER_ABSTIME flag.
```
I don't think we need to expose C implementation details -- from an API usage perspective, I'd rather we call these arguments "interval" and "initial_expiration", and make them keyword-only.

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,205 @@ os_times_impl(PyObject *module)
+    it_interval_ns: long_long = 0
+        similar to timerfd_settime except for in nanoseconds.
+    it_value_ns: long_long = 0
+        similar to timerfd_settime except for in nanoseconds.
```
See comments on ``timerfd_settime``. Please also copy the full docstrings here rather than saying read X instead.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,168 @@ features:
+   If the timerfd counter is zero and :const:`TFD_NONBLOCK` is not
+   specified, :func:`read` blocks.
+
+   Example::
```
These two examples feel out-of-place in the documentation for `timerfd_create` specifically. How about adding a new subheading for timer file descriptions (to include all 5 functions / 4 flags) and put the examples in the introduction to this new subsection?

AA-Turner replied:
You marked this "resolved", but the examples remain in-line. Is there a comment I missed?

m-tmatma replied:
@AA-Turner 
Sorry, I don't remember there is time gap.

> These two examples feel out-of-place in the documentation for timerfd_create specifically. How about adding a new subheading for timer file descriptions (to include all 5 functions / 4 flags) and put the examples in the introduction to this new subsection?

I don't have enough knowledge on rst.
Could you suggest it concretely?

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,168 @@ features:
+   .. versionadded:: 3.13
+
+
+.. function:: timerfd_settime(fd, flags, interval, value)
```
It feels odd to have ``interval`` before ``value`` here -- given ``interval`` can be 0 and the timer still works, but if ``value`` is 0 the timer is turned off.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,168 @@ features:
+
+   :c:expr:`interval = new_value.it_interval.tv_sec + new_value.it_interval.tv_nsec * 1e-9`
+
+   :c:expr:`value = new_value.it_value.tv_sec + new_value.it_value.tv_nsec * 1e-9`
```
This behaviour is surprising. I'd expect that using the nanosecond version of the function overrides any previous values, and vice-versa with the float version of the function -- but it seems their values are combined. I think users should be **warned** of this behaviour.

m-tmatma replied:
@AA-Turner 
Added comment at 3d55e2ecff.

AA-Turner replied:
I think it is much better to change the suprising behaviour if possible rather than adding a comment. However if you disagree with changing the behaviour, we should rephrase and make this a warning to the user.

```rest
.. caution::
   The seconds set by this function and the nanoseconds set by 
   :func:`timerfd_settime_ns` are summed by the timer descriptor, which may cause surpising results.
   For example, setting an interval of 2 seconds via this function and 2**9 seconds via :func:`timerfd_settime_ns` will make the actual interval of the timer file descriptor **4 seconds**, not 2.
```

m-tmatma replied:
@AA-Turner 
What do you mean?
I'm not sure what is your concerns.

Setting an interval of 2 seconds via `timerfd_settime` will make the time file descriptor in every  2 seconds.

Setting an interval of 2**9 seconds(you mean 512 seconds?) via `timerfd_settime_ns` will make the time file descriptor in every 512 seconds.

And the original comment no longer exists on the lastest branch.

AA-Turner committed:
Documentation review

AA-Turner reviewed (commented):
On API usage, I think all the functions should have *fd* as positional only and the other arguments as kwonly, eg ``def timerfd_settime(fd, /, *, flags, interval, initial_expiration)``. I also think *flags* should default to 0 at least for timer creation.

What happens if a negative number is passed to settime? This should be documented.

I also think we should try to abstract away from C a little more, the manpage can talk about the gory details.

I've pushed a commit with an intitial docs review, feel free to revert / ask on any parts!

A

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+
+.. function:: timerfd_gettime(fd, /)
+
+   Return a two-item tuple (``interval``, ``value``).
```
It's odd that this tuple is (interval, value) -- I'd expect the other way around, as the next expiration is more important than the interval.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+
+   .. versionadded:: 3.13
+
+.. data:: TFD_NONBLOCK
```
All four flags should explain what the *effect* is, not just that they set some attribute at the C level.

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/whatsnew/3.13.rst:
```diff
@@ -143,6 +143,16 @@ opcode
+
+* Add :func:`os.timerfd_create`, :func:`os.timerfd_settime`, :func:`os.timerfd_gettime`, :func:`os.timerfd_settime_ns`,
+  and :func:`os.timerfd_gettime_ns` to provide a low level interface for Linux's timer notification file descriptor.
+  (Contributed by Masaru Tsuchiyama in :gh:`108277`.)
```
Put the main object first (the interface to the timerfd system)
```suggestion
* Add a low level interface for Linux's timer notification file descriptors
  via :func:`os.timerfd_create`,
  :func:`os.timerfd_settime`, :func:`os.timerfd_settime_ns`,
  :func:`os.timerfd_gettime`, and :func:`os.timerfd_gettime_ns`.
  (Contributed by Masaru Tsuchiyama in :gh:`108277`.)
```

AA-Turner commented on Doc/whatsnew/3.13.rst:
```diff
@@ -143,6 +143,16 @@ opcode
+  (Contributed by Masaru Tsuchiyama in :gh:`108277`.)
+
+* Add :const:`os.TFD_NONBLOCK`, :const:`os.TFD_CLOEXEC`, :const:`os.TFD_TIMER_ABSTIME`, and :const:`os.TFD_TIMER_CANCEL_ON_SET`
+  (Contributed by Masaru Tsuchiyama in :gh:`108277`.)
```
Are these distinct from the timerfd interface?

m-tmatma replied:
@AA-Turner 
This is a part of the timerfd interface.
Should I remove?

m-tmatma replied:
@AA-Turner 
Merged the sections.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks.
+
+   The following example shows how to use a timer file descriptor
+   to execute a function twice a second:
```
Would this run forever (until an exception)? This should be made explicit

m-tmatma replied:
@AA-Turner 
Changed the sample code.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   (set by *clockid* in :func:`timerfd_create`) reaches *value* seconds.
+
+   Return a two-item tuple of (``interval``, ``value``) from
+   the previous timer state, before this function executed.
```
Same point that the tuple order seems odd

m-tmatma replied:
@AA-Turner 
Reorderd.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   By default timer will fire when *value* seconds have elapsed.
+   However, if the :const:`TFD_TIMER_ABSTIME` flag is set,
+   the timer will fire when the timer's clock
+   (set by *clockid* in :func:`timerfd_create`) reaches *value* seconds.
```
Is there any way to query the internal clock given an arbitrary timerfd?

m-tmatma replied:
@AA-Turner 
I think there is no such operation.

AA-Turner replied:
Should we add one?

m-tmatma replied:
@AA-Turner 
You mean we should save the internal clock inside and return the saved value if requested?

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,6 +3780,174 @@ features:
+   The file descriptor supports :func:`read` with a buffer size of 8,
+   :func:`~select.select`, and :func:`~select.poll`.
+
+   .. seealso:: The :manpage:`timerfd_create(2)` man page.
```
The positioning of the ``seealso`` directives is inconsistent -- some are after the intro para and some are at the end.

m-tmatma replied:
@AA-Turner 
Moved.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10035,6 +10040,207 @@ os_times_impl(PyObject *module)
+        it_interval is interval for timer in seconds.
+    it_value: double = 0.0
+        it_value is initial expiration timing in seconds. It could be absolute time or relative time
+        based on TFD_TIMER_ABSTIME flag.
```
This feels quite difficult to read. Two suggestions: remove the ``it_`` prefix, and turn this into a bullet point list with a prose overview. You also have the key info "it_interval is interval for timer in seconds." at the very end!

m-tmatma replied:
@AA-Turner 
Revised.

AA-Turner reviewed (commented):
A few bits I missed.

Continuing the design questions, is there any way to query the flags of a particular timerfd, or its clock ID?

A

m-tmatma committed:
Update Doc/whatsnew/3.13.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Fix error on runing Tools/clinic/clinic.py -f Modules/posixmodule.c

$ python3 Tools/clinic/clinic.py -f Modules/posixmodule.c
Error in file 'Modules/posixmodule.c' on line 10120:
Docstring for 'os.timerfd_gettime' does not have a summary line!
Every non-blank function docstring must start with a single line summary followed by an empty line.

m-tmatma committed:
apply the result of 'python3 Tools/clinic/clinic.py -f Modules/posixmodule.c'

m-tmatma committed:
rename valut to initial_expiration

m-tmatma committed:
rename return value

m-tmatma committed:
reverse the order of set/get functions

m-tmatma committed:
merge sections for new functions and new consts

m-tmatma committed:
fix error check

m-tmatma committed:
reorder parameters and tuple of return values

m-tmatma committed:
reorder function definition

m-tmatma committed:
move seealso to the end of section

m-tmatma committed:
move code example

m-tmatma committed:
reorder initialize to parameters

m-tmatma committed:
change parameter to kw only except for fd

m-tmatma committed:
fix document

m-tmatma committed:
fix document

m-tmatma committed:
update document

m-tmatma committed:
check the value from read()

m-tmatma committed:
fix lint error

m-tmatma committed:
refer to negative values

m-tmatma committed:
add type comment

m-tmatma committed:
reorder.

m-tmatma committed:
fix parameter

m-tmatma committed:
add supported functions

m-tmatma committed:
rename parameter

m-tmatma committed:
fix comment

m-tmatma committed:
fix const comment

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
update comment on os.timerfd_settime_ns

m-tmatma committed:
Fix document

m-tmatma committed:
fix lint error

m-tmatma committed:
add test params

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,16 +3780,43 @@ features:
 
-   .. seealso:: The :manpage:`timerfd_create(2)` man page.
+   The file descriptor returned by :func:`timerfd_create` supports the following
+   additional operations:
```
```suggestion
   The file descriptor returned by :func:`timerfd_create` supports:
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,16 +3780,43 @@ features:
+   If the timer has already expired one or more times, :func:`read` returns
+   the number of expirations in :class:`bytes` of host byte order. The return value
+   can be coverted to :class:`int` value by :func:`int.from_bytes` with ``byteorder``
+   is :const:`sys.byteorder`.
```
Thank you for adding more detail! 

What happens if I call ``fd.read()`` and it hasn't expired yet?

```suggestion
   The file descriptor supports :func:`read` with a buffer size of 8.
   If the timer has already expired one or more times, :func:`read` returns
   the number of expirations with the host's endianness,
   which may be coverted to an :class:`int` by ``int.from_bytes(x, byteorder=sys.byteorder)``.
```

This section might be better converted to a recipie in the new subsection I mentioned earlier. Another question: Can I call ``.read()`` multiple times? On a normal file descriptor, at some point I'll reach the end of the file.

m-tmatma replied:
@AA-Turner 
Updated.


If there is no timer event after the previous `.read()` call, it blocks until new timer event.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,16 +3780,43 @@ features:
+   can be coverted to :class:`int` value by :func:`int.from_bytes` with ``byteorder``
+   is :const:`sys.byteorder`.
+
+   :func:`~select.select` and :func:`~select.poll` can be used to wait until
+   timer expires and the file descriptor is readable.
```
Again I think a recipie should be provided

m-tmatma replied:
@AA-Turner 
Added examples.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3780,16 +3780,43 @@ features:
    *clockid* must be a valid :ref:`clock ID <time-clock-id-constants>`,
-   as defined in the :py:mod:`time` module.
+   as defined in the :py:mod:`time` module and the following values are
+   available.
```
```suggestion
   as defined in the :py:mod:`time` module:
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3801,55 +3828,19 @@ features:
-      interval = 0.5  # Set the timer interval to 0.5 seconds
-      value = 1  # Start the timer in 1 second
-      os.timerfd_settime(fd, 0, interval, value)  # Start the timer
+   If :const:`TFD_CLOEXEC` is set as a flag, set close-on-exec flag for new file descriptor.
```
Should we explain the close-on-exec flag?

m-tmatma replied:
@AA-Turner 
There are some similar consts.
But there are no additional documents.

```
$ git grep CLOEXEC | grep PyModule_AddIntMacro
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, F_DUPFD_CLOEXEC)) return -1;
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, F_DUP2FD_CLOEXEC)) return -1;
Modules/fcntlmodule.c:    if (PyModule_AddIntMacro(m, FD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, O_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, TFD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, MFD_CLOEXEC)) return -1;
Modules/posixmodule.c:    if (PyModule_AddIntMacro(m, EFD_CLOEXEC)) return -1;

```

m-tmatma replied:
@AA-Turner 
**_CLOEXEC should be documented, 
but I think it's out of the scope of this PR.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
    - :const:`TFD_TIMER_ABSTIME`
    - :const:`TFD_TIMER_CANCEL_ON_SET`
 
+   The timer is disabled by setting *initial* :py:class:`float` to zero (``0``).
```
```suggestion
   The timer is disabled by setting *initial* to zero (``0``).
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+
+      import os, time
+
+      fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)
```
Given flags now defaults to 0:

```suggestion
      fd = os.timerfd_create(time.CLOCK_REALTIME)
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+      fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)
+      initial_expiration = 1  # Start the timer in 1 second
+      interval = 0.5  # Set the timer interval to 0.5 seconds
+      os.timerfd_settime(fd, flags=0, initial=initial_expiration, interval=interval)  # Start the timer
```
Similarly:
```suggestion
      os.timerfd_settime(fd, initial=initial_expiration, interval=interval)  # Start the timer
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+
+      import os, time
+
+      fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)
```
```suggestion
      fd = os.timerfd_create(time.CLOCK_REALTIME)
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+      fd = os.timerfd_create(time.CLOCK_REALTIME, flags=0)
+      initial_expiration = 10**9  # Start the timer in 1 second
+      interval = 10**9 // 2  # Set the timer interval to 0.5 seconds
+      os.timerfd_settime_ns(fd, flags=0, initial=initial_expiration, interval=interval)  # Start the timer
```
```suggestion
      os.timerfd_settime_ns(fd, initial=initial_expiration, interval=interval)  # Start the timer
```

m-tmatma replied:
@AA-Turner 
When I tried that, I got the following error.
How can you do that?

```
======================================================================
ERROR: test_timerfd_ns_interval (test.test_os.TimerfdTests.test_timerfd_ns_interval)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/..../cpython/Lib/test/test_os.py", line 4096, in test_timerfd_ns_interval
    _, _ = os.timerfd_settime_ns(fd, initial=initial_expiration_ns, interval=interval_ns)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: timerfd_settime_ns() missing required argument 'flags' (pos 2)
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3801,55 +3828,19 @@ features:
 
 
-.. function:: timerfd_settime(fd, flags, interval=0.0, value=0.0, /)
+.. function:: timerfd_settime(fd, /, *, flags=flags, initial=0.0, interval=0.0)
```
``flags`` isn't a default!
```suggestion
.. function:: timerfd_settime(fd, /, *, flags=0, initial=0.0, interval=0.0)
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
 .. function:: timerfd_gettime(fd, /)
 
-   Return a two-item tuple (``interval``, ``value``).
+   Return a two-item tuple (``next_expiration``, ``interval``).
```
```suggestion
   Return a two-item tuple of floats (``next_expiration``, ``interval``).
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3917,6 +3962,7 @@ features:
 
    A flag for the :func:`timerfd_create` function,
    which sets the :const:`O_NONBLOCK` status flag for the new timer file descriptor.
+   :func:`read` to the new timer file descriptor will not be blocked.
```
This doesn't read right grammar-wise

m-tmatma replied:
@AA-Turner 
Fixed.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3934,14 +3980,17 @@ features:
 .. data:: TFD_TIMER_ABSTIME
 
    A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions.
+   Use as an absolute value on the timer's clock.
```
This doesn't make sense

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3934,14 +3980,17 @@ features:
-   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions.
+   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions
+   along with :const:`TFD_TIMER_ABSTIME`. Timer is aborted when real-time clock is
+   changed discontinuously.
```
```suggestion
   A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions
   along with :const:`TFD_TIMER_ABSTIME`.
   The timer is cancelled when the time of the underlying clock changes discontinuously
   (for example, during a daylight-savings time transition).
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
 build_itimerspec(const struct itimerspec* curr_value)
 {
+    double _next_expiration = (double)(curr_value->it_value.tv_sec) +
+                    (double)curr_value->it_value.tv_nsec * 1e-9;
```
Could this magic number be extracted out?

m-tmatma replied:
@AA-Turner 
Fixed.

AA-Turner replied:
Thanks!

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
+        return NULL;
+    }
     double _interval = (double)(curr_value->it_interval.tv_sec) +
                        (double)curr_value->it_interval.tv_nsec * 1e-9;
```
Ditto

m-tmatma replied:
@AA-Turner 
Fixed.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
 
     clockid: int
-        CLOCK_REALTIME or CLOCK_MONOTONIC. CLOCK_REALTIME.
+        CLOCK_REALTIME or CLOCK_MONOTONIC.
```
```suggestion
        A valid clock ID constant.
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
+        CLOCK_REALTIME
+            A settable system-wide real-time clock.
+            If system clock is changed, timer setting need to be updated.
+            See TFD_TIMER_ABSTIME and TFD_TIMER_CANCEL_ON_SET.
```
```suggestion
            See os.TFD_TIMER_ABSTIME and os.TFD_TIMER_CANCEL_ON_SET.
```

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
-        CLOCK_REALTIME or CLOCK_MONOTONIC. CLOCK_REALTIME.
+        CLOCK_REALTIME or CLOCK_MONOTONIC.
+
+        CLOCK_REALTIME
```
```suggestion
        time.CLOCK_REALTIME
```

m-tmatma replied:
@AA-Turner
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
+            If system clock is changed, timer setting need to be updated.
+            See TFD_TIMER_ABSTIME and TFD_TIMER_CANCEL_ON_SET.
+
+        CLOCK_MONOTONIC
```
```suggestion
        time.CLOCK_MONOTONIC
```

m-tmatma replied:
@AA-Turner
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10058,50 +10058,83 @@ os_times_impl(PyObject *module)
+           Even if system clock is changed, timer setting will be not affected.
+
+        CLOCK_BOOTTIME
+           Same as CLOCK_MONOTONIC except that it inclues any time that
```
```suggestion
        time.CLOCK_BOOTTIME
           Same as time.CLOCK_MONOTONIC except that it inclues any time that
```

m-tmatma replied:
@AA-Turner
Updated.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10114,90 +10147,112 @@ os_timerfd_create_impl(PyObject *module, int clockid, int flags)
+    flags: int
+        0 or a bit mask of TFD_TIMER_ABSTIME or TFD_TIMER_CANCEL_ON_SET.
+    initial: double = 0.0
+        initial expiration timing in seconds.
```
```suggestion
        The initial expiration time, in seconds.
```

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10114,90 +10147,112 @@ os_timerfd_create_impl(PyObject *module, int clockid, int flags)
+        * If *flags* has TFD_TIMER_ABSTIME bit and TFD_TIMER_CANCEL_ON_SET bit and *flags* and time.CLOCK_REALTIME
+          and system clock is changed discontinuously, reading a file descriptor is aborted with ECANCELED.
+    interval: double = 0.0
+        interval for the timer in seconds.
```
```suggestion
        The timer's interval, in seconds.
```

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10114,90 +10147,112 @@ os_timerfd_create_impl(PyObject *module, int clockid, int flags)
-        based on TFD_TIMER_ABSTIME flag.
-    /
+    initial: long_long = 0
+        initial expiration timing in seconds.
```
same changes as to timerfd_settime

m-tmatma replied:
@AA-Turner 
Updated.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+   The timer is disabled by setting *initial* :py:class:`float` to zero (``0``).
+   If *initial* is equal to or greater than zero, the timer is enabled.
+   If *initial* is less than zero, it raises :class:`OSError` with ``errno``
+   set to :const:`errno.EINVAL`
```
```suggestion
   If *initial* is less than zero, it raises an :class:`OSError` exception
   with ``errno`` set to :const:`errno.EINVAL`
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3859,44 +3850,98 @@ features:
+   If *initial* is less than zero, it raises :class:`OSError` with ``errno``
+   set to :const:`errno.EINVAL`
+
+   By default timer will fire when *initial* seconds have elapsed.
```
```suggestion
   By default the timer will fire when *initial* seconds have elapsed.
```

AA-Turner reviewed (commented):
Thank you for making the comprehensive changes! A re-review:

A

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
fix typo

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
update by Tools/clinic/clinic.py

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
fix document

m-tmatma committed:
divide to a new subsection

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Add comment.

m-tmatma committed:
add select.epoll test case

m-tmatma committed:
make flags positional argument and its default value to 0

m-tmatma committed:
fix timeout

m-tmatma committed:
add tests for select.epoll() and os.timerfd_settime_ns

m-tmatma committed:
make non-blocking for select, epoll tests

m-tmatma committed:
fix assert

m-tmatma committed:
cleanup epoll

m-tmatma committed:
rename variable of epoll

m-tmatma committed:
fix code samples

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
move example

m-tmatma committed:
mention errno.EAGAIN

m-tmatma committed:
change to useful sample

m-tmatma committed:
fix comment

m-tmatma committed:
update example

m-tmatma committed:
fix indentation

m-tmatma committed:
Add comment

m-tmatma committed:
define a macro to convert seconds in double.

m-tmatma committed:
update comment

m-tmatma committed:
update comment

m-tmatma committed:
update comment

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
add comment

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
add comment

m-tmatma committed:
fix code sample

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
remove log

m-tmatma committed:
reorder

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
add reference

m-tmatma committed:
move decription

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3784,26 +3784,109 @@ features:
 .. function:: timerfd_create(clockid, /, *, flags)
 
    Create and return a timer file descriptor (*timerfd*).
+   The file descriptor must be closed with :func:`os.close` on cleanup.
```
I'd move this later in the overview of the function. Also "cleanup" should be clarified -- perhps "to properly clean up the underlying resources" or similar.

m-tmatma replied:
@AA-Turner 
Fixed.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3784,26 +3784,109 @@ features:
 
-   The file descriptor returned by :func:`timerfd_create` supports the following
-   additional operations:
+   The file descriptor returned by :func:`timerfd_create` supports:
```
Could you explain the rationale for splitting this up? I liked the previous version (below), but I'm open to arguments.

Previous version (with endianness update):

```rest
The file descriptor returned by :func:`timerfd_create` supports the following
additional operations:

   - :func:`read`
   - :func:`~select.select`
   - :func:`~select.poll`.

   The file descriptor supports :func:`read` with a buffer size of 8.
   If the timer has already expired one or more times, :func:`read` returns
   the number of expirations with the host's endianness,
   which may be converted to an :class:`int` by ``int.from_bytes(x, byteorder=sys.byteorder)``.

   :func:`~select.select` and :func:`~select.poll` can be used to wait until
   timer expires and the file descriptor is readable.
```

m-tmatma replied:
@AA-Turner 
I couldn't find where, but I adapted your suggestion.

m-tmatma replied:
@AA-Turner 
Thank you for pushing.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3826,8 +3909,9 @@ features:
-   :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks.
+   If :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks until the timer expires.
+   If it is set as a flag, :func:`read` doesn't block, but if there is no timer expiration,
+   :func:`read` raises :class:`OSError` with ``errno`` is set to :const:`errno.EAGAIN`.
```
"no timer expiration" -- does this mean if the timer *won't* expire, or if there hasn't been an expiration since the last call to ``read``?

m-tmatma replied:
@AA-Turner 
Yes. Exactly.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3838,9 +3922,10 @@ features:
 
 
-.. function:: timerfd_settime(fd, /, *, flags=flags, initial=0.0, interval=0.0)
+.. function:: timerfd_settime(fd, flags=0, /, *, initial=0.0, interval=0.0)
```
Why did we revert to having flags pos-only?
```suggestion
.. function:: timerfd_settime(fd, /, *, flags=0, initial=0.0, interval=0.0)
```

m-tmatma replied:
@AA-Turner 
I changed keyword arg, but as I said earlier, I coudn't make it dual arg for pos/keywork.
If you can, please share the concrete way or push the branch directly.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3875,28 +3960,34 @@ features:
    is aborted with with the error ECANCELED.
 
+   Linux manages system clock as UTC. A daylight-savings time transition is done
+   by changing time offset only and doesn't cause discontinuous system clock change.
```
Can we provide an example of what *would* cause a discontinuous system clock change? I'm unsure if we should list something that does not cause a discontinuous change.

m-tmatma replied:
@AA-Turner 

https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> if the real-time clock undergoes a discontinuous change
              ([settimeofday(2)](https://man7.org/linux/man-pages/man2/settimeofday.2.html), [clock_settime(2)](https://man7.org/linux/man-pages/man2/clock_settime.2.html), or similar).  When
              such changes occur, a current or future [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) from the
              file descriptor will fail with the error ECANCELED.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3875,28 +3960,34 @@ features:
    to execute a function twice a second:
 
-   .. code:: python
+   Example::
```
We should prefer ``.. code:: python`` over `::`. Also note that the text above says "the following example", so the "Example:" header isn't needed here

m-tmatma replied:
@AA-Turner 
But there is no `.. code:: python` anywhere in the python code tree.
Are you suggesting new comment style here?

To avoid miscommunication, please suggest concrete code.

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3905,29 +3996,33 @@ features:
+.. function:: timerfd_settime_ns(fd, flags=0, /, *, initial=0, interval=0)
 
    Similar to :func:`timerfd_settime`, but use time as nanoseconds.
+   This function operates the same interval timer as :func:`timerfd_settime`.
```
Same notes as above

m-tmatma replied:
@AA-Turner 
Updated.
(https://github.com/python/cpython/pull/108382#discussion_r1334960739)

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3981,7 +4076,8 @@ features:
    A flag for the :func:`timerfd_settime` and :func:`timerfd_settime_ns` functions.
-   Use as an absolute value on the timer's clock.
+   If this flag is set, the functions use as an absolute value on the timer's clock
+   in UTC not in local time.
```
I think this means the same thing:

```suggestion
   If this flag is set, *initial* is interpreted as an absolute value on the timer's clock
   (in UTC seconds or nanoseconds since the Unix Epoch).
```

AA-Turner commented on Doc/library/os.rst:
```diff
@@ -3990,8 +4086,8 @@ features:
-   along with :const:`TFD_TIMER_ABSTIME`. Timer is aborted when real-time clock is
-   changed discontinuously.
+   along with :const:`TFD_TIMER_ABSTIME`.
+   The timer is cancelled when the time of the underlying clock changes discontinuously.
```
This is clearer, thanks! Same note as above -- please could we provide an example of a discontinuous change?

m-tmatma replied:
@AA-Turner 
See 
https://github.com/python/cpython/pull/108382#discussion_r1334995659


https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> if the real-time clock undergoes a discontinuous change
              ([settimeofday(2)](https://man7.org/linux/man-pages/man2/settimeofday.2.html), [clock_settime(2)](https://man7.org/linux/man-pages/man2/clock_settime.2.html), or similar).  When
              such changes occur, a current or future [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) from the
              file descriptor will fail with the error ECANCELED.

AA-Turner commented on Modules/posixmodule.c:
```diff
@@ -10072,18 +10072,19 @@ os_times_impl(PyObject *module)
 #if defined(HAVE_TIMERFD_API)
 #define ONE_SECOND_IN_NS (1000 * 1000 * 1000)
 #define EXTRACT_NSEC(value)  (long)( ( (double)(value) - (time_t)(value) ) * 1e9)
+#define CONVERT_SEC_AND_NSEC_TO_DOUBLE(sec, nsec) ( (double)(sec) + (double)(nsec) * 1e-9 )
```
Thanks!

AA-Turner reviewed (commented):
Thanks for making the changes! Another re-review

A

m-tmatma committed:
Update Doc/library/os.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

m-tmatma committed:
move and fix the comment of cleanup

m-tmatma committed:
make flags a keyword parameter

AA-Turner committed:
Add a subsection for timerfd

AA-Turner committed:
Rephrase closing

AA-Turner committed:
Merge select into the read list, rephrase read

m-tmatma commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,304 @@ features:
    .. versionadded:: 3.10
 
 
+Timer File Descriptors
```
@AA-Turner 
Thank you for pushing.

m-tmatma commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,304 @@ features:
+Timer File Descriptors
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 3.3
```
@AA-Turner 
typo for 3.13?

AA-Turner replied:
Oops!

AA-Turner committed:
3.3 -> 3.13

m-tmatma committed:
Explain the trigger of system clock change.

m-tmatma committed:
Explain error on non-blocking fd.

m-tmatma committed:
add description

m-tmatma committed:
fix sample

m-tmatma committed:
update sample

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+        time.CLOCK_REALTIME
+            A settable system-wide real-time clock.
+            If system clock is changed, timer setting need to be updated.
+            See os.TFD_TIMER_ABSTIME and os.TFD_TIMER_CANCEL_ON_SET.
```
I do not think that it should be in docstring. You can add it in the documentation, but even then it should be described rather at the module level than in this function description. The man page contains shorter description.

m-tmatma replied:
@serhiy-storchaka 
Removed.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+    /
+    *
+    flags: int = 0
+        0 or a bit mask of os.TFD_NONBLOCK or os.TFD_CLOEXEC.
```
It is enough. The rest should be in the documentation.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+
+        * If *flags* has TFD_TIMER_ABSTIME bit, *flags* must be in absolute time.
+        * If *flags* doesn't have TFD_TIMER_ABSTIME bit, *flags* must be in relative time.
+        * If *flags* has TFD_TIMER_ABSTIME bit and TFD_TIMER_CANCEL_ON_SET bit and *flags* and time.CLOCK_REALTIME
```
This line is 115 characters long. And some more indentation will be added in the `help(os)` output. According to PEP 8, the limit for docstrings is 72 characters.

In any case, this description is too long and verbose. Details, such as what happen if system clock is changed discontinuously, can be left for the documentation. Some statements should be merged and shortened (more use "else" or "otherwise" instead of repeating).

m-tmatma replied:
@serhiy-storchaka 
Removed.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+    struct itimerspec old_value;
+    int result;
+    new_value.it_value.tv_sec = (time_t)initial;
+    new_value.it_value.tv_nsec = EXTRACT_NSEC(initial);
```
cc @vstinner

It may be vulnerable to overflow and loss of precision. There may be a special C API to set `timespec` struct. For example `_PyTime_AsTimespec()`/`_PyTime_AsTimespec_clamp()`, but how to get `_PyTime_t`?

vstinner replied:
The _PyTime_t API can be used here, right. You should add a _PyTime_FromSecondsDouble() function (move code from pytime_from_object, float code path) and then use _PyTime_AsTimespec().

Rounding is hard :-( Also, PyTime API check for NaN, your code doesn't.

m-tmatma replied:
@vstinner
Fixed.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+
+
+/*[clinic input]
+os.timerfd_settime_ns
```
@vstinner: What is better, add separate function with the `_ns` suffix, or add parameters like `interval_ns`, or accept not only float but some more complex object for precise time interval?

vstinner replied:
The status quo is to have 2 APIs: https://peps.python.org/pep-0564/

* Python float: number of seconds
* Python int, ``_ns`` variant: integer number of seconds

Like time.time() returns a float and time.time_ns() returns an int.

m-tmatma replied:
@serhiy-storchaka @vstinner 
OK?

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+    struct itimerspec old_value;
+    int result;
+    new_value.it_value.tv_sec = initial / ONE_SECOND_IN_NS;
+    new_value.it_value.tv_nsec = initial % ONE_SECOND_IN_NS;
```
Perhaps `_PyTime_FromNanoseconds()` or `_PyTime_FromNanosecondsObject()` followed by `_PyTime_AsTimespec()`/`_PyTime_AsTimespec_clamp()` could be used here to avoid integer overflow.

cc @vstinner

vstinner replied:
Yeah, please use the _PyTime_t API which takes care of rounding, overflow errors, etc.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+build_itimerspec(const struct itimerspec* curr_value)
+{
+    double _next_expiration = CONVERT_SEC_AND_NSEC_TO_DOUBLE(curr_value->it_value.tv_sec,
+                                                          curr_value->it_value.tv_nsec);
```
`_PyTime_FromTimespec()` followed by `_PyTime_AsSecondsDouble()` can be used here.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+static PyObject *
+build_itimerspec(const struct itimerspec* curr_value)
+{
+    double _next_expiration = CONVERT_SEC_AND_NSEC_TO_DOUBLE(curr_value->it_value.tv_sec,
```
`next_expiration` is too long name. Why not use simple `value`?

m-tmatma replied:
@serhiy-storchaka 
`value` is too common to understand the meaning.
So I chose the word by its meaning.

m-tmatma replied:
@serhiy-storchaka 
Changed to value because this is just an untility.
We don't need to know the acutual meaning.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+{
+    long long _next_expiration = (long long)curr_value->it_value.tv_sec * ONE_SECOND_IN_NS +
+                    (long long)curr_value->it_value.tv_nsec;
+    PyObject *next_expiration = PyLong_FromLongLong(_next_expiration);
```
`_PyTime_FromTimespec()` followed by `_PyTime_AsNanosecondsObject()` can be used here.

m-tmatma replied:
@serhiy-storchaka 
changed to use _PyTime_FromTimespec

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+        Py_DECREF(next_expiration);
+        return NULL;
+    }
+    PyObject *tuple = PyTuple_Pack(2, next_expiration, interval);
```
The code can be simplified by using `Py_BuildValue("dd", ...)` instead of `PyTuple_Pack(2, ...)`. But it is slower, and I do not know whether this sloweness matters, so I left it up to you.

m-tmatma replied:
@serhiy-storchaka 
If it's slower, I'll keep it as it is.

serhiy-storchaka commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,425 @@ features:
+       for _ in range(4):
+          # read() will block until the timer expires
+          _ = os.read(fd, 8)
+          print("Timer expired")
```
It looks to me like 3-space indentation instead of PEP 8 recommended 4-space indentation.

m-tmatma replied:
@serhiy-storchaka 
I'll fix it.

The indent of `.. code-block:: python` is 3-space.
So I'm confused.

m-tmatma replied:
@serhiy-storchaka 
Fixed the indent at f06e3525b3d1ae2f59d0adfba00279d398e3fbd3.

serhiy-storchaka commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,425 @@ features:
+   - :const:`TFD_NONBLOCK`
+   - :const:`TFD_CLOEXEC`
+
+   If :const:`TFD_NONBLOCK` is not set as a flag, :func:`read` blocks until the timer expires.
```
Too long lines. It makes reviewing difficult. I do not remember the exact limit for documentation, but it is not more than 80 columns.

m-tmatma replied:
@serhiy-storchaka 
Fixed.

serhiy-storchaka reviewed (commented):
In general LGTM. There may be some issues with integer overflow and loss of precision, so it is better to use existing helpers. @vstinner can suggest more.

Please make docstrings shorter. Details should go in the documentation, docstring should only describe what the function do and what parameters mean. Also reformat them to fit in 72 characters limit.

Examples in the documentation are too long. The `os` module documentation is already large, and it makes reading it and searching necessary information more difficult. Could them be moved in the "How To" file or in separate Python files?

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
fix indent

m-tmatma committed:
limit less than 80 chars in a line

m-tmatma committed:
add comment

vstinner commented on Lib/test/test_os.py:
```diff
@@ -3922,6 +3922,360 @@ def test_eventfd_select(self):
+
+    def test_timerfd_non_blocking(self):
+        size = 8  # read 8 bytes
+        fd = os.timerfd_create(time.CLOCK_REALTIME, flags=os.TFD_NONBLOCK)
```
Would it be possible to test that the FD is actually non-blocking? Maybe using fcntl?

vstinner commented on Lib/test/test_os.py:
```diff
@@ -3922,6 +3922,360 @@ def test_eventfd_select(self):
+class TimerfdTests(unittest.TestCase):
+    def test_timerfd_initval(self):
+        fd = os.timerfd_create(time.CLOCK_REALTIME)
+        self.assertNotEqual(fd, -1)
```
I would expect this check to be in the C code. If you prefer to check also the FD here, maybe add a helper function:

```py
    def timerfd_create(self, *args):
        fd = os.timerfd_create(*args)
        self.assertGreaterEqual(fd, 0)
        return fd
```

Please test that FD is non-inheritable (see my comment below): os.get_inheritable(fd) must be False.

m-tmatma replied:
@vstinner 
Fixed.

vstinner commented on Lib/test/test_os.py:
```diff
@@ -3922,6 +3922,360 @@ def test_eventfd_select(self):
+        self.assertNotEqual(fd, -1)
+        self.addCleanup(os.close, fd)
+
+        initial_expiration = 0.25
```
What is the resolution of these timers? I'm curious how you chose place=3. Please add a comment to document the expected resolution.

vstinner commented on Modules/posixmodule.c:
```diff
@@ -10060,6 +10065,252 @@ os_times_impl(PyObject *module)
+{
+    int fd;
+    Py_BEGIN_ALLOW_THREADS
+    fd = timerfd_create(clockid, flags);
```
[PEP 446](https://peps.python.org/pep-0446/) requires that all FDs created by Python are non-inheritable. Please always pass TFD_CLOEXEC:
```suggestion
    flags |= TFD_CLOEXEC;  // PEP 446: always create non-inheritable FD
    fd = timerfd_create(clockid, flags);
```

m-tmatma replied:
@vstinner 
Applied your suggestion.

vstinner commented on Modules/posixmodule.c:
```diff
@@ -16699,6 +16968,10 @@ static const struct have_function {
 #endif
 
+#ifdef HAVE_TIMERFD_API
+    {"HAVE_TIMERFD_API", NULL},
```
Is it really useful? Why not just checking for hasattr(os, "timerfd_create")? I suggest removing it.

m-tmatma replied:
@vstinner 
Could you suggest a concrete code?
Or you can push it directly.

vstinner replied:
I suggest to remove this constant. That's my suggestion.

m-tmatma replied:
@vstinner 
Only Linux suppports timerfd API.
Is it OK to just remove it?

vstinner replied:
You didn't reply to my question. Why not testing ``hasattr(os, "timerfd_create")`` instead of testing `os.HAVE_TIMERFD_API`?

m-tmatma replied:
@vstinner 
timerfd API has timerfd_create, and timerfd_setttime, timerfd_gettime.
So I create the macro.

Replaced it with HAVE_TIMERFD_CREATE.
Do this resolve your concerns?

vstinner commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,435 @@ features:
+   import os, time
+
+   # Create the timer file descriptor
+   fd = os.timerfd_create(time.CLOCK_REALTIME)
```
I dislike this clock, it can jump backward/forward at sysadmin will. Why using using MONOTONIC for examples?

m-tmatma replied:
@vstinner 
1. linux manual uses it as an example.
https://man7.org/linux/man-pages/man2/timerfd_create.2.html

> it can jump backward/forward at sysadmin will

2. Users can detect system clock change by calling timerfd_settime with TFD_TIMER_ABSTIME and TFD_TIMER_CANCEL_ON_SET. If system clock changes, timer will be canceled.
Users need to choose right flag based on the purpose.

vstinner replied:
TFD_TIMER_CANCEL_ON_SET is really nice.

m-tmatma replied:
@vstinner 
Resolved?

vstinner commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,435 @@ features:
+   # Start the timer with os.timerfd_settime_ns() in nanoseconds.
+   # Timer 1 fires every 0.25 seconds; timer 2 every 0.5 seconds; etc
+   for i, fd in enumerate(fds, start=1):
+       i = i * 10**9
```
10**9 looks mysterious, please add a constant and give it a name ;-) For example, ONE_SECOND.

m-tmatma replied:
@vstinner 
Fixed.

vstinner commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,435 @@ features:
+   timeout = 3
+   try:
+       conn = None
+       isActive = True
```
I suggest renaming the variable to is_active.

m-tmatma replied:
@vstinner 
Fixed.

vstinner commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,435 @@ features:
+       # Register the timer file descriptor for read events
+       ep.register(fd, select.EPOLLIN)
+
+   # In this example, use loopback address to send "stop" command to the server.
```
Maybe move it before the creation of the fds, it's more likely to fail if the port is already used.

m-tmatma replied:
@vstinner 
Moved.

vstinner commented on Doc/library/os.rst:
```diff
@@ -3781,6 +3781,435 @@ features:
+   :const:`errno.EAGAIN`.
+
+   If :const:`TFD_CLOEXEC` is set as a flag, the close-on-exec flag is set
+   for the new file descriptor.
```
You should explain that it's always set by Python.

m-tmatma replied:
@vstinner 
Fixed.

m-tmatma committed:
Update Modules/posixmodule.c

Co-authored-by: Victor Stinner <vstinner@python.org>

m-tmatma committed:
Use const

m-tmatma committed:
Confirm non-blocking fd

m-tmatma committed:
define an utility function

m-tmatma committed:
rename variable

m-tmatma committed:
check if fd is not inheritable

m-tmatma committed:
reorder socket creation

m-tmatma committed:
fix comment

m-tmatma committed:
create _PyTime_FromSecondsDouble and use _PyTime_AsTimespec

m-tmatma committed:
use _PyTime_AsTimespec

vstinner commented on Modules/posixmodule.c:
```diff
@@ -10060,6 +10065,249 @@ os_times_impl(PyObject *module)
+    struct itimerspec old_value;
+    int result;
+    _PyTime_AsTimespec(_PyTime_FromSecondsDouble(initial), &new_value.it_value);
+    _PyTime_AsTimespec(_PyTime_FromSecondsDouble(interval), &new_value.it_interval);
```
You must not ignore silently overflow errors: you can write "if (as_timespec() < 0) return NULL;``.

Same in os_timerfd_settime_ns_impl().

m-tmatma replied:
@vstinner 
Fixed

vstinner commented on Python/pytime.c:
```diff
@@ -427,6 +427,15 @@ _PyTime_FromSeconds(int seconds)
+_PyTime_t
+_PyTime_FromSecondsDouble(double seconds)
+{
+    _PyTime_t t = (_PyTime_t)(seconds * SEC_TO_NS);
```
Rounding is undefined here. You should reuse pytime_from_double() with SEC_TO_NS, and use the right rounding mode in your code. time.clock_settime() uses _PyTime_ROUND_FLOOR rounding mode.

m-tmatma replied:
@vstinner 
Fixed.

m-tmatma committed:
use _PyTime_FromTimespec and Py_BuildValue

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
use pytime_from_double to implement _PyTime_FromSecondsDouble

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
add error check

m-tmatma committed:
fix error handling

vstinner reviewed (commented):
Please try to fix warnings in posixmodule.c, like:

```
 Check warning on line 10104 in Modules/posixmodule.c
GitHub Actions
/ Ubuntu

passing argument 2 of ‘_PyTime_FromTimespec’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
```

They can be seen at https://github.com/python/cpython/pull/108382/files

m-tmatma committed:
rename variable

m-tmatma committed:
fix warnings

m-tmatma committed:
remove extra docstring

m-tmatma committed:
move samples to HOWTO

vstinner commented on Lib/test/test_os.py:
```diff
@@ -3922,6 +3922,366 @@ def test_eventfd_select(self):
+    def timerfd_create(self, *args, **kwargs):
+        fd = os.timerfd_create(*args, **kwargs)
+        self.assertGreaterEqual(fd, 0)
+        self.assertFalse(os.get_inheritable(fd))
```
It seems like self.addCleanup(os.close, fd) calls can be moved there.

m-tmatma replied:
@vstinner 
Fixed.

vstinner commented on Python/pytime.c:
```diff
@@ -635,6 +635,16 @@ _PyTime_AsNanosecondsObject(_PyTime_t t)
 }
 
+_PyTime_t
+_PyTime_FromSecondsDouble(double seconds)
```
I would prefer to pass rounding mode as a function argument, rather than hardcode the rounding to _PyTime_ROUND_FLOOR.

m-tmatma replied:
@vstinner 
Fixed.

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
move common code to a utility function

m-tmatma committed:
add rounding parameter to _PyTime_FromSecondsDouble

m-tmatma committed:
Merge branch 'main' into feature/os.timerfd_xxx

m-tmatma committed:
replace HAVE_TIMERFD_API with HAVE_TIMERFD_CREATE

vstinner reviewed (approved):
LGTM. Thanks for all updates!

IMO using _PyTime_t API and passing the rounding method from the time module is the right way to handle rounding.
